Face rec

schelian opened the pull request:
Using RosPack to find directories, etc.

schelian committed:
security use case (load .pickle file, etc.)

schelian committed:
using rospkg to set dirs

Commit:
updating paths, removing some demo code

athackst commented on src/openface_node.py:
```diff
@@ -18,7 +18,23 @@
 
+rp = rospkg.RosPack()
+_OPENFACE_DIR = rp.get_path('openface_ros') # find openface (should be at same level as openface_ros per install instructions)
+_OPENFACE_DIR = _OPENFACE_DIR[:-4] + "/" 
```
Let's make this a parameter instead.  There is no reason to force someone to put openface in a certain place.

athackst commented on src/openface_node.py:
```diff
@@ -18,7 +18,23 @@
+  print "demo lib directory does not exist; exiting"
+  sys.exit()
+
+_FACE_DICT_FNAME = 'toyota_faces.pickle'
```
This is too specific for a general ros package.  Please convert to parameter

schelian replied:
OK

athackst commented on src/openface_node.py:
```diff
@@ -76,13 +92,17 @@ def __init__(self, align_path, net_path, storage_folder):
 
+        with open( _DEMO_LIB_DIR + _FACE_DICT_FNAME, 'rb') as f:
+            self._face_dict = pickle.load( f )
+            print "read _face_dict: " + _DEMO_LIB_DIR + _FACE_DICT_FNAME
```
convert to rospy.loginfo or similar so that this will get logged.

athackst commented on src/openface_node.py:
```diff
@@ -92,11 +112,16 @@ def _update_detections_with_recognitions(self, detections):
         detections = [self._update_detection_with_recognition(d) for d in detections]
 
         # Now find the detection index with highest name probability, only for image in /tmp/faces
+        min_distance = 1000.
```
It looks like this isn't used

athackst commented on src/openface_node.py:
```diff
@@ -92,11 +112,16 @@ def _update_detections_with_recognitions(self, detections):
+                # this code properlly labels bounding boxes but can caused other problems in the code...
+                #if ( l2_distances[min_index] < min_distance ):
+                #    min_distance = l2_distances[min_index];
+                #    detections[min_index]["name"] = name
```
Decide whether or not this is needed and remove if not.

athackst commented on src/openface_node.py:
```diff
@@ -166,13 +191,19 @@ def _learn_face_srv(self, req):
+        # from http://www.diveintopython3.net/serializing.html
+        with open( _DEMO_LIB_DIR + _FACE_DICT_FNAME, 'wb' ) as f:
+            pickle.dump( self._face_dict, f );
+            print "wrote _face_dict: " + _DEMO_LIB_DIR + _FACE_DICT_FNAME
```
convert to rospy.log message

athackst commented on src/openface_node.py:
```diff
@@ -166,13 +191,19 @@ def _learn_face_srv(self, req):
         cv2.imwrite("%s/%s_detect.jpeg" % (self._storage_folder, now.strftime("%Y-%m-%d-%H-%M-%d-%f")), bgr_image)
 
         for d in detections:
+            #print d
```
remove commented code or convert to rospy.debug so we can decide if it gets printed at runtime.

athackst commented on src/openface_node.py:
```diff
@@ -181,6 +212,9 @@ def _save_images(self, detections, bgr_image):
                 if "name" in d:
+                    #print "*****"
+                    #print d["name"]
+                    #print "*****"
```
removed commented code or convert to rospy.debug

athackst commented on src/openface_node.py:
```diff
@@ -76,13 +92,17 @@ def __init__(self, align_path, net_path, storage_folder):
-
         self._face_dict = {}  # Mapping from string to list of reps
 
+        with open( _DEMO_LIB_DIR + _FACE_DICT_FNAME, 'rb') as f:
```
this is forcing the user to have a pickle file, which is changing functionality.  Is it possible to make this optional?

athackst commented on src/openface_node.py:
```diff
@@ -166,13 +191,19 @@ def _learn_face_srv(self, req):
         rospy.loginfo("Succesfully learned face of '%s'" % req.name)
 
+        # from http://www.diveintopython3.net/serializing.html
+        with open( _DEMO_LIB_DIR + _FACE_DICT_FNAME, 'wb' ) as f:
```
This is potentially forcing a change in functionality.  Consider making optional

athackst commented on src/openface_node.py:
```diff
@@ -213,15 +247,16 @@ def _detect_face_srv(self, req):
-        rospy.logerr("This is for debugging")
-        rospy.logerr(detections)
+        #rospy.logerr("This is for debugging")
+        #rospy.logerr(detections)
```
Why were these removed?

schelian replied:
Put a comment in the code but basically the original version did too much logging.

athackst commented on src/openface_node.py:
```diff
@@ -18,7 +18,23 @@
+  sys.exit()
+
+_FACE_DICT_FNAME = 'toyota_faces.pickle'
+_SAVE = False
```
This should be a parameter and it looks like the default for it should be "True" to maintain previous functionality.  We can launch with it set to "False" to maintain your logic.

schelian replied:
Ok, now using rosparam a lot; similarly prints are now rospy.loginfo's.

Re: the pickle file, if no pickle file is passed in, we're back to the original logic.

I also removed dead / commented out code.

athackst commented on src/openface_node.py:
```diff
@@ -236,10 +271,17 @@ def _detect_face_srv(self, req):
+    if ( not camera_lr ):
+        camera_lr = "_" + camera_lr
+
+    rospy.init_node('openface' + camera_lr)
```
rospy.init_node should happen before any parameters are retrieved.

schelian replied:
In this case, I was trying to have 2 different nodes, one for camera left and another for camera right.  I can remove this since the triangulation function is not fully supported yet.

athackst replied:
That still doesn't make any sense.  If you want to have 2 nodes, you make 2 nodes.  The init_node name only looks hardcoded.  You can assign the name at runtime, either through a roslaunch file name=".." or command line __name:=".."

athackst commented on src/openface_node.py:
```diff
@@ -236,10 +271,17 @@ def _detect_face_srv(self, req):
+
+    rospy.init_node('openface' + camera_lr)
+
+    align_path_param = rospy.get_param('~align_path', _OPENFACE_DIR + 'models/dlib/shape_predictor_68_face_landmarks.dat')
```
all of these global variables would be better served as parameters loaded in here.

athackst commented on test/test_learn_detect.py:
```diff
@@ -9,16 +9,31 @@
 
-rospy.init_node('test_learn_detect')
+camera_lr = rospy.get_param("~camera_lr","");
+print "camera_lr: " + camera_lr
```
convert to ros.log message

athackst commented on test/test_learn_detect.py:
```diff
@@ -41,18 +71,80 @@ def callback(data):
     elif key == 1048676 or key == 100: # D
-        print detect_srv(image=data, external_api_request=external_api_request)
+        detections = detect_srv(image=data, external_api_request=external_api_request)
+        #print detections
```
remove or change to rospy.logdebug

athackst commented on test/test_learn_detect.py:
```diff
@@ -41,18 +71,80 @@ def callback(data):
+            disp_str = "(no detections)"
+            if ( camera_lr ):
+                disp_str = camera_lr + ": " + disp_str
+            print disp_str
```
change to rospy.log message

athackst commented on test/test_learn_detect.py:
```diff
@@ -41,18 +71,80 @@ def callback(data):
+
+        if ( detections.face_detections.__len__() == 0):
+            disp_str = "(no detections)"
+            if ( camera_lr ):
```
You don't need this if statement.  Log messages should be happening in the function calls below

athackst commented on test/test_learn_detect.py:
```diff
@@ -41,18 +71,80 @@ def callback(data):
+            print disp_str
+        else:
+            match_name = '(unknown)'
+            # occasionally, the first catch of no detections will not catch so make sure a distances really has something
```
explain this.

schelian replied:
the comment is wrong here.  it should say, there is a detection but it's of an unknown person.

athackst commented on test/test_learn_detect.py:
```diff
@@ -41,18 +71,80 @@ def callback(data):
+            disp_str = names[ min_distance_idx ] + ' with distance ' + str( min_distance )
+            if ( camera_lr ):
+                disp_str = camera_lr + ": " + disp_str
+            print disp_str
```
change to rospy.log

athackst commented on test/test_learn_detect.py:
```diff
@@ -41,18 +71,80 @@ def callback(data):
+                min_distance_idx = idx
+        #print distances
+        #print min_distance
+        #print min_distance_idx
```
remove or make debug message

athackst commented on test/test_learn_detect.py:
```diff
@@ -41,18 +71,80 @@ def callback(data):
+# subscribes to images with its callback (its callback learns/detects, calling the server)
 image_sub = rospy.Subscriber("image", Image, callback)
 rospy.loginfo("Listening to %s -- spinning .." % image_sub.name)
 rospy.loginfo("Usage: L to learn, D to detect")
```
The "test" folder should be reserved for unit testing.  However since this was here before, I will allow it.

athackst commented on test/triangulate.py:
```diff
@@ -0,0 +1,75 @@
+#!/usr/bin/env python
```
since this file doesn't work as intended, I would remove it and place it in a separate branch.

Commit:
adding dependency to openface; adjusting paths

Commit:
changing from prints to rospy.log...; removing assumption that openface is a catkin pkg (it isn't yet), moving things to rosparam's, etc.

removing test/triangulate.py (not ready for prime time; should just use Asus depth data anyway)

schelian reviewed (commented):
Ok, now using rosparam a lot; similarly prints are now rospy.loginfo's.

Re: the pickle file, if no pickle file is passed in, we're back to the original logic.

I also removed dead / commented out code.

athackst commented on src/openface_node.py:
```diff
@@ -213,15 +229,17 @@ def _detect_face_srv(self, req):
-        rospy.logerr("This is for debugging")
-        rospy.logerr(detections)
+        # leave these commented out; otherwise things are too verbose
+        #rospy.logerr("This is for debugging")
```
You should change to a rospy.logdebug(..) not just comment it out.

There's a reason there's a logger level.

schelian replied:
Ok

athackst commented on src/openface_node.py:
```diff
@@ -96,7 +105,8 @@ def _update_detections_with_recognitions(self, detections):
                 l2_distances = [ dict(zip(d["names"], d["l2_distances"]))[name] for d in detections ]
                 min_index = l2_distances.index(min(l2_distances))
-                detections[min_index]["name"] = name
+                detections[min_index]["name"] = name # @bug
```
This sounds like a pretty big bug, no?

schelian replied:
Yes but I fix the results in the detections case of test_learn_detect.py

Commit:
removing camera_lr stuff; logdebug of detections

schelian commented:
Hi @athackst and @brandonnorthcutt-tri, can you please submit your final comments?  I'd like to move on to the other things.
